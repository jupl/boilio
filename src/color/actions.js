import {createAction} from 'redux-actions'

/**
 * Create action that to move to the next color
 * @type {Function}
 */
export const previousColor = createAction('COLORS_PREVIOUS_COLOR')

/**
 * Create action that to move to the previous color
 * @type {Function}
 */
export const nextColor = createAction('COLORS_NEXT_COLOR')

/**
 * Create action that to move to the next color automatically
 * @type {Function}
 */
export const autoNextColor = createAction('COLORS_AUTO_NEXT_COLOR')
