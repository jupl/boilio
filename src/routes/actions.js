import {createAction} from 'redux-actions'

/** Create action that to remove current scene */
export const popRoute = createAction('ROUTES_POP')

/** Create action that to switch to a scene */
export const pushRoute = createAction('ROUTES_PUSH', push)

/** Create action to update contents of a scene */
export const refreshRoute = createAction('ROUTES_REFRESH', refresh)

/**
 * Build action data from params when pushRoute action is called
 * @param {string} key - Key identifier for a scene
 * @param {?Object} params - Param data to pass over to scene
 * @return {Object} Action data
 */
function push(key, params) {
  return params != null ? {key, params} : {key}
}

/**
 * Build action data from params when refreshRoute action is called
 * @param {Object} params - Param data to pass over to scene
 * @return {Object} Action data
 */
function refresh(params) {
  return params != null ? {params} : {}
}
