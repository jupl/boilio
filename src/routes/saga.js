import {takeEvery} from 'redux-saga'
import {Actions} from 'react-native-router-flux'
import {ACTION_TYPES} from './constants'

/**
 * Start process to listen in on route actions
 * @return {void}
 */
export default function* saga() {
  yield [
    takeEvery(ACTION_TYPES.POP_ROUTE, () => Actions.pop()),
    takeEvery(ACTION_TYPES.PUSH_ROUTE, pushRoute),
    takeEvery(ACTION_TYPES.REFRESH, refreshRoute),
  ]
}

/**
 * Transition to another scene with given scene key and params
 * @param {Object} action - Action emitted by Redux dispatcher
 * @param {Object} action.payload - Action data
 * @param {Object} action.payload.key - Key identifier for a scene
 * @param {?Object} action.payload.params - Optional data to pass to scene
 * @return {void}
 */
function pushRoute({payload: {key, params}}) {
  Actions[key](params != null ? {params} : {})
}

/**
 * Update curren scene with given params
 * @param {Object} action - Action emitted by Redux dispatcher
 * @param {Object} action.payload - Action data
 * @param {?Object} action.payload.params - Optional data to pass to scene
 * @return {void}
 */
function refreshRoute({payload: {params}}) {
  Actions.refresh(params != null ? {params} : {})
}
