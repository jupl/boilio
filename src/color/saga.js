import {delay} from 'redux-saga'
import {call, put, race, take} from 'redux-saga/effects'
import {previousColor, nextColor, autoNextColor} from './actions'

/**
 * Start process to cycle colors automatically until an action is emitted to
 * manually change color
 * @return {void}
 */
export default function* saga() {
  for(;;) {
    const {end} = yield race({
      timeout: call(delay, 3000),
      end: take([`${previousColor}`, `${nextColor}`]),
    })
    if(end) { break }
    yield put(autoNextColor())
  }
}
