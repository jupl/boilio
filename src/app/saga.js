import {call} from 'redux-saga/effects'
import color from '../color/saga'
import routes from '../routes/saga'

// Collection of all saga tasks to initiate
const sagas = [color, routes]

/**
 * Entry point to start running all initial saga tasks
 * @return {void}
 */
export default function* saga() {
  yield sagas.map(call)
}
