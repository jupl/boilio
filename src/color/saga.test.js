import saga from './saga'
import {nextColor} from './actions'

describe('Color saga', () => {
  it('should cycle until a previous/next event is fired', () => {
    // Start off saga
    const it = saga()
    expect(it.next()).toMatchSnapshot()

    // Ensure that auto next color event is emitted by default
    Array(2).fill().forEach(() => {
      expect(it.next({})).toMatchSnapshot()
      expect(it.next()).toMatchSnapshot()
    })

    // Verify saga ends when end is reached
    expect(it.next({end: nextColor()})).toMatchSnapshot()
  })
})
