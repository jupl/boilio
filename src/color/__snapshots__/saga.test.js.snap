exports[`Color saga should cycle until a previous/next event is fired 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "RACE": Object {
      "end": Object {
        "@@redux-saga/IO": true,
        "TAKE": Object {
          "pattern": Array [
            "COLORS_PREVIOUS_COLOR",
            "COLORS_NEXT_COLOR",
          ],
        },
      },
      "timeout": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            3000,
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
  },
}
`;

exports[`Color saga should cycle until a previous/next event is fired 2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "COLORS_AUTO_NEXT_COLOR",
      },
      "channel": null,
    },
  },
}
`;

exports[`Color saga should cycle until a previous/next event is fired 3`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "RACE": Object {
      "end": Object {
        "@@redux-saga/IO": true,
        "TAKE": Object {
          "pattern": Array [
            "COLORS_PREVIOUS_COLOR",
            "COLORS_NEXT_COLOR",
          ],
        },
      },
      "timeout": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            3000,
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
  },
}
`;

exports[`Color saga should cycle until a previous/next event is fired 4`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "PUT": Object {
      "action": Object {
        "type": "COLORS_AUTO_NEXT_COLOR",
      },
      "channel": null,
    },
  },
}
`;

exports[`Color saga should cycle until a previous/next event is fired 5`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "RACE": Object {
      "end": Object {
        "@@redux-saga/IO": true,
        "TAKE": Object {
          "pattern": Array [
            "COLORS_PREVIOUS_COLOR",
            "COLORS_NEXT_COLOR",
          ],
        },
      },
      "timeout": Object {
        "@@redux-saga/IO": true,
        "CALL": Object {
          "args": Array [
            3000,
          ],
          "context": null,
          "fn": [Function],
        },
      },
    },
  },
}
`;

exports[`Color saga should cycle until a previous/next event is fired 6`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
